### PR TITLE
En 5965 update validators pub key when epoch change and add rating information

### PR DESCRIPTION
**Pull request by miiu96**

### CODE REVIEW

File: core/errors.go [hunk 1]
```diff
@@ -10,8 +10,8 @@ var ErrNilMarshalizer = errors.New("nil marshalizer provided")
 // ErrNilHasher signals that a nil hasher has been provided
 var ErrNilHasher = errors.New("nil hasher provided")
 
-// ErrNilCoordinator signals that a nil shardCoordinator has been provided
-var ErrNilCoordinator = errors.New("nil coordinator provided")
+// ErrNilNodesCoordinator signals that an operation has been attempted to or with a nil nodes coordinator
```
Comment (bogdan-rosianu) [inline core/errors.go:13]: please change this comment.

File: core/indexer/common.go [hunk 2]
```diff
@@ -252,3 +252,69 @@ func buildReceiptTransaction(
...
+					"tx hash", mb.Hash)
+				continue
+			}
+		} else {
+			// update miniblock
+			meta = []byte(fmt.Sprintf(`{ "update" : { "_id" : "%s" } }%s`, mb.Hash, "\n"))
+			serializedData = []byte(fmt.Sprintf(`{ "doc": { "receiverBlockHash" : "%s" } }`, mb.ReceiverBlockHash))
+		}
+
+		// append a newline foreach element
```
Comment (bogdan-rosianu) [inline core/indexer/common.go:278]: for each *

File: core/indexer/common.go [hunk 3]
```diff
@@ -252,3 +252,69 @@ func buildReceiptTransaction(
...
+		_, err = buff.Write(serializedData)
+		if err != nil {
+			log.Warn("elastic search: serialize bulk miniblocks, write serialized miniblock", "error", err.Error())
+		}
+	}
+
+	return buff
+}
+
+func serializeBulkTx(bulk []*Transaction) bytes.Buffer {
```
Comment (bogdan-rosianu) [inline core/indexer/common.go:294]: Txs ?

File: core/indexer/common.go [hunk 4]
```diff
@@ -252,3 +252,69 @@ func buildReceiptTransaction(
...
+		var err error
+		var meta, serializedData []byte
+		if hdrShardID == mb.SenderShardID {
+			//insert miniblock
+			meta = []byte(fmt.Sprintf(`{ "index" : { "_id" : "%s", "_type" : "%s" } }%s`, mb.Hash, "_doc", "\n"))
+			serializedData, err = json.Marshal(mb)
+			if err != nil {
+				log.Debug("indexer: marshal",
+					"error", "could not serialize miniblock, will skip indexing",
+					"tx hash", mb.Hash)
```
Comment (bogdan-rosianu) [inline core/indexer/common.go:269]: mb hash

File: core/indexer/common.go [hunk 5]
```diff
@@ -252,3 +252,69 @@ func buildReceiptTransaction(
...
+	for _, tx := range bulk {
+		meta := []byte(fmt.Sprintf(`{ "index" : { "_id" : "%s", "_type" : "%s" } }%s`, tx.Hash, "_doc", "\n"))
+		serializedTx, err := json.Marshal(tx)
+		if err != nil {
+			log.Debug("indexer: marshal",
+				"error", "could not serialize transaction, will skip indexing",
+				"tx hash", tx.Hash)
+			continue
+		}
+		// append a newline foreach element
```
Comment (bogdan-rosianu) [inline core/indexer/common.go:305]: for each

File: core/indexer/elasticsearch.go [hunk 6]
```diff
@@ -86,43 +96,57 @@ func (ei *elasticIndexer) SaveBlock(
...
 	if len(body.MiniBlocks) == 0 {
 		log.Debug("indexer", "error", ErrNoMiniblocks.Error())
 		return
 	}
 
-	if ei.options.TxIndexingEnabled {
-		go ei.database.SaveTransactions(body, headerHandler, txPool, ei.shardCoordinator.SelfId())
-	}
-}
+	go ei.database.SaveMiniblocks(headerHandler, body)
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearch.go:106]: would it worth to add also a flag condition for indexing miniblocks ? maybe add a TODO or wait for the second/third reviewer

File: core/indexer/elasticsearch.go [hunk 7]
```diff
@@ -86,43 +96,57 @@ func (ei *elasticIndexer) SaveBlock(
...
 }
 
-//SaveValidatorsPubKeys will send all validators public keys to elastic search
-func (ei *elasticIndexer) SaveValidatorsPubKeys(validatorsPubKeys map[uint32][][]byte) {
+func (ei *elasticIndexer) epochStartEventHandler() epochStart.EpochStartHandler {
+	subscribeHandler := notifier.NewHandlerForEpochStart(func(hdr data.HeaderHandler) {
+		currentEpoch := hdr.GetEpoch()
+		validatorsPubKeys, err := ei.coordinator.GetAllEligibleValidatorsPublicKeys(currentEpoch)
+		if err != nil {
+			log.Warn("GetAllEligibleValidatorPublicKeys for epoch 0 failed", "error", err)
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearch.go:123]: is not epoch 0 anymore, please change log message

File: core/indexer/elasticsearchDatabase.go [hunk 8]
```diff
@@ -85,14 +86,24 @@ func (esd *elasticSearchDatabase) createIndexes() error {
...
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // SaveHeader will prepare and save information about a header in elasticsearch server
-func (esd *elasticSearchDatabase) SaveHeader(header data.HeaderHandler, signersIndexes []uint64) {
+func (esd *elasticSearchDatabase) SaveHeader(header data.HeaderHandler, signersIndexes []uint64, body *block.Body, notarizedHeadersHashes []string) {
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearchDatabase.go:103]: move parameters on separate lines

File: core/indexer/elasticsearchDatabase.go [hunk 9]
```diff
@@ -213,32 +243,60 @@ func (esd *elasticSearchDatabase) buildTransactionBulks(
...
+		return
+	}
+	if len(miniblocks) == 0 {
+		return
+	}
+
+	buff := serializeBulkMiniBlocks(header.GetShardID(), miniblocks)
+	err := esd.dbWriter.DoBulkRequest(&buff, miniblocksIndex)
+	if err != nil {
+		log.Warn("indexer", "error1", "indexing bulk of miniblocks",
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearchDatabase.go:260]: why not `log.Warn("indexing bulk of miniblock", "error", err)` ?

File: core/indexer/elasticsearchDatabase.go [hunk 10]
```diff
@@ -303,6 +361,38 @@ func (esd *elasticSearchDatabase) SaveShardValidatorsPubKeys(shardId uint32, sha
 	}
 }
 
+// SaveValidatorsRating will save validators rating
+func (esd *elasticSearchDatabase) SaveValidatorsRating(Index string, validatorsRatingInfo []ValidatorRatingInfo) {
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearchDatabase.go:365]: `index` instead of `Index`

File: core/indexer/elasticsearch_test.go [hunk 11]
```diff
@@ -36,10 +36,10 @@ func NewElasticIndexerArguments() indexer.ElasticIndexerArgs {
 		Url:              "url",
 		UserName:         "user",
 		Password:         "password",
-		ShardCoordinator: &mock.ShardCoordinatorMock{},
 		Marshalizer:      &mock.MarshalizerMock{},
 		Hasher:           &mock.HasherMock{},
 		Options:          &indexer.Options{},
+		NodesCoordinator: &mock.NodesCoordinatorMock{},
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearch_test.go:42]: adapt the mock after merge. there is a compile issue right now

File: core/indexer/elasticsearch_test.go [hunk 12]
```diff
@@ -269,7 +241,8 @@ func TestElasticIndexer_SaveValidatorsPubKeys(t *testing.T) {
 
 	keys := [][]byte{[]byte("key")}
 	valPubKey[0] = keys
-	ei.SaveValidatorsPubKeys(valPubKey)
+	epoch := uint32(0)
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearch_test.go:244]: add a test for epoch start notifier and you can also check if the ES is updated when a new epoch starts

File: core/errors.go (hunk 1 above)
Comment (miiu96) [reply core/errors.go:13]: done.

File: core/indexer/common.go (hunk 2 above)
Comment (miiu96) [reply core/indexer/common.go:278]: done.

File: core/indexer/common.go (hunk 3 above)
Comment (miiu96) [reply core/indexer/common.go:294]: done.

File: core/indexer/common.go (hunk 4 above)
Comment (miiu96) [reply core/indexer/common.go:269]: done.

File: core/indexer/common.go (hunk 5 above)
Comment (miiu96) [reply core/indexer/common.go:305]: done.

File: core/indexer/elasticsearch.go (hunk 7 above)
Comment (miiu96) [reply core/indexer/elasticsearch.go:123]: done.

File: core/indexer/elasticsearchDatabase.go (hunk 8 above)
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:103]: done.

File: core/indexer/elasticsearchDatabase.go (hunk 9 above)
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:260]: done.  refactored

File: core/indexer/elasticsearchDatabase.go (hunk 10 above)
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:365]: done.

File: core/indexer/elasticsearch_test.go (hunk 12 above)
Comment (miiu96) [reply core/indexer/elasticsearch_test.go:244]: done.

File: core/indexer/elasticsearch.go [hunk 13]
```diff
@@ -120,7 +120,7 @@ func (ei *elasticIndexer) epochStartEventHandler() epochStart.EpochStartHandler
 		currentEpoch := hdr.GetEpoch()
 		validatorsPubKeys, err := ei.coordinator.GetAllEligibleValidatorsPublicKeys(currentEpoch)
 		if err != nil {
```
Comment (bogdan-rosianu) [inline core/indexer/elasticsearch.go:122]: Log also the epoch

Comment (miiu96) [reply core/indexer/elasticsearch.go:122]: done.

File: core/indexer/elasticsearchDatabase.go [hunk 14]
```diff
@@ -213,32 +259,59 @@ func (esd *elasticSearchDatabase) buildTransactionBulks(
...
+			SenderShardID:   miniblock.SenderShardID,
+			ReceiverShardID: miniblock.ReceiverShardID,
+			Type:            miniblock.Type.String(),
 		}
-		_, err = buff.Write(serializedTx)
-		if err != nil {
-			log.Warn("elastic search: serialize bulk tx, write serialized tx", "error", err.Error())
+
+		if mb.SenderShardID == header.GetShardID() {
+			mb.SenderBlockHash = encodedHeaderHash
```
Comment (raduchis) [inline core/indexer/elasticsearchDatabase.go:306]: Is this done to make some links between headers and miniblocks to show in the explorer?

File: core/indexer/elasticsearchDatabase_test.go [hunk 15]
```diff
@@ -193,7 +210,7 @@ func TestElasticsearch_saveShardValidatorsPubKeys_RequestError(t *testing.T) {
 		},
 	}
 	elasticDatabase := newTestElasticSearchDatabase(dbWriter, arguments)
-	elasticDatabase.SaveShardValidatorsPubKeys(shardId, valPubKeys)
+	elasticDatabase.SaveShardValidatorsPubKeys(shardId, epoch, valPubKeys)
```
Comment (raduchis) [inline core/indexer/elasticsearchDatabase_test.go:213]: Could you please add another test with 2 epochs to see that different lists of validatorPubKeys are requested if different epochs are provided?

File: core/indexer/elasticsearch_test.go [hunk 16]
```diff
@@ -53,15 +56,6 @@ func TestElasticIndexer_NewIndexerWithNilUrlShouldError(t *testing.T) {
 	require.Equal(t, core.ErrNilUrl, err)
 }
 
-func TestElasticIndexer_NewIndexerWithNilShardCoordinatorShouldError(t *testing.T) {
```
Comment (raduchis) [inline core/indexer/elasticsearch_test.go:56]: there are no tests for the newly added NodesCoordinator and EpochStartNotifier.

File: core/indexer/elasticsearch_test.go [hunk 17]
```diff
@@ -230,51 +225,67 @@ func TestElasticIndexer_SaveMetaBlockNilHeader(t *testing.T) {
...
-	valPubKey := make(map[uint32][][]byte)
+	arguments.ShardId = core.MetachainShardId
+	epochChangeNotifier := &mock.EpochStartNotifierStub{}
+	arguments.EpochStartNotifier = epochChangeNotifier
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	arguments.NodesCoordinator = &mock.NodesCoordinatorMock{
+		GetAllEligibleValidatorsPublicKeysCalled: func() (m map[uint32][][]byte, err error) {
```
Comment (raduchis) [inline core/indexer/elasticsearch_test.go:277]: the epoch is not actually used. Could you please add the epoch to the header and also to the GetAllEligibleValidatorsPublicKeysCalled in order to validate that the NotifyAll was called with the correct epoch.

File: process/block/metablock.go [hunk 18]
```diff
@@ -523,9 +525,20 @@ func (mp *metaProcessor) indexBlock(
 		return
 	}
 
-	go mp.core.Indexer().SaveBlock(body, metaBlock, txPool, signersIndexes)
+	go mp.core.Indexer().SaveBlock(body, metaBlock, txPool, signersIndexes, notarizedHeadersHashes)
 
-	saveRoundInfoInElastic(mp.core.Indexer(), mp.nodesCoordinator, core.MetachainShardId, metaBlock, lastMetaBlock, signersIndexes)
+	indexRoundInfo(mp.core.Indexer(), mp.nodesCoordinator, core.MetachainShardId, metaBlock, lastMetaBlock, signersIndexes)
+
+	if metaBlock.GetNonce() == 1 {
```
Comment (raduchis) [inline process/block/metablock.go:532]: is there a way to merge the 2 ifs: this one and the one below?

File: process/block/metrics.go [hunk 19]
```diff
@@ -226,7 +228,38 @@ func saveRoundInfoInElastic(
...
+	if err != nil {
+		return
+	}
+
+	validators, err := valStatProc.GetValidatorInfoForRootHash(latestHash)
+	if err != nil {
+		return
+	}
+
+	for shardID, validatorInfosInShard := range validators {
```
Comment (raduchis) [inline process/block/metrics.go:251]: please add a todo to use the validatorInfoProvider GetLatestValidators() method in order to have the rating already computed and not divide by 1.000.000 this is actually the maxRating

File: cmd/node/main.go [hunk 20]
```diff
@@ -1061,13 +1063,14 @@ func indexValidatorsListIfNeeded(elasticIndexer indexer.Indexer, coordinator sha
 		return
 	}
 
-	validatorsPubKeys, err := coordinator.GetAllEligibleValidatorsPublicKeys(0)
+	epoch := uint32(0)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go:1066]: done.

File: core/indexer/elasticsearchDatabase.go (hunk 14 above)
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:306]: yes. We want to link miniblocks in explorer to see when a miniblocks was included in sender shard and  also in receiver shard.

File: core/indexer/elasticsearch_test.go (hunk 16 above)
Comment (miiu96) [reply core/indexer/elasticsearch_test.go:56]: done.

File: core/indexer/elasticsearch_test.go (hunk 17 above)
Comment (miiu96) [reply core/indexer/elasticsearch_test.go:277]: done.

File: process/block/metablock.go (hunk 18 above)
Comment (miiu96) [reply process/block/metablock.go:532]:
```
if metaBlock.GetNonce() != 1 && !metaBlock.IsStartOfEpochBlock(){
		return
}
```

done.

File: process/block/metrics.go (hunk 19 above)
Comment (miiu96) [reply process/block/metrics.go:251]: done.

File: core/indexer/elasticsearchDatabase.go [hunk 21]
```diff
@@ -114,27 +130,46 @@ func (esd *elasticSearchDatabase) SaveHeader(header data.HeaderHandler, signersI
...
+		Round:                 header.GetRound(),
+		Epoch:                 header.GetEpoch(),
+		ShardID:               header.GetShardID(),
+		Hash:                  hex.EncodeToString(headerHash),
+		MiniBlocksHashes:      miniblocksHashes,
+		NotarizedBlocksHashes: notarizedHeadersHashes,
+		Proposer:              signersIndexes[0],
+		Validators:            signersIndexes,
+		PubKeyBitmap:          hex.EncodeToString(header.GetPubKeysBitmap()),
+		Size:                  int64(len(h)),
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:168]: i added a todo for that and will be fixed on a future pull request

File: core/indexer/elasticsearchDatabase_test.go (hunk 15 above)
Comment (miiu96) [reply core/indexer/elasticsearchDatabase_test.go:213]: validators list is requested at a upper level in elasticIndexer so there i will add a test with 2 epoch.

File: core/indexer/elasticsearchDatabase.go (hunk 14 above)
Comment (raduchis) [reply core/indexer/elasticsearchDatabase.go:307]: ok

File: core/indexer/elasticsearch_test.go (hunk 16 above)
Comment (raduchis) [reply core/indexer/elasticsearch_test.go:56]: ok

File: cmd/node/main.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/main.go]: ok

File: core/indexer/elasticsearchDatabase_test.go (hunk 15 above)
Comment (raduchis) [reply core/indexer/elasticsearchDatabase_test.go:213]: ok

File: core/indexer/elasticsearchDatabase.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply core/indexer/elasticsearchDatabase.go]: ok

File: core/indexer/elasticsearch_test.go (hunk 17 above)
Comment (raduchis) [reply core/indexer/elasticsearch_test.go]: ok

File: process/block/metablock.go (hunk 18 above)
Comment (raduchis) [reply process/block/metablock.go]: ok

File: process/block/metrics.go (hunk 19 above)
Comment (raduchis) [reply process/block/metrics.go:251]: ok

**Pull request merged**